# This Notebook will be mainly used for the Coursera_Capstone 

In [1]:
import pandas as pd
import numpy as np

In [2]:
print("Hello Capstone Project Course! ")

Hello Capstone Project Course! 


# Segmenting and Clustering Neighborhoods in Toronto

## scraping data from wikipedia

In [3]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

res = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
soup = BeautifulSoup(res.content,'lxml')
table = soup.find_all('table')[0]
df = pd.read_html(str(table))
print(df[0].to_json(orient='records'))

[{"Postal Code":"M1A","Borough":"Not assigned","Neighborhood":null},{"Postal Code":"M2A","Borough":"Not assigned","Neighborhood":null},{"Postal Code":"M3A","Borough":"North York","Neighborhood":"Parkwoods"},{"Postal Code":"M4A","Borough":"North York","Neighborhood":"Victoria Village"},{"Postal Code":"M5A","Borough":"Downtown Toronto","Neighborhood":"Regent Park, Harbourfront"},{"Postal Code":"M6A","Borough":"North York","Neighborhood":"Lawrence Manor, Lawrence Heights"},{"Postal Code":"M7A","Borough":"Downtown Toronto","Neighborhood":"Queen's Park, Ontario Provincial Government"},{"Postal Code":"M8A","Borough":"Not assigned","Neighborhood":null},{"Postal Code":"M9A","Borough":"Etobicoke","Neighborhood":"Islington Avenue"},{"Postal Code":"M1B","Borough":"Scarborough","Neighborhood":"Malvern, Rouge"},{"Postal Code":"M2B","Borough":"Not assigned","Neighborhood":null},{"Postal Code":"M3B","Borough":"North York","Neighborhood":"Don Mills"},{"Postal Code":"M4B","Borough":"East York","Neighbo

In [4]:
type(df)

list

In [5]:
type(df[0])

pandas.core.frame.DataFrame

In [6]:
#assigning df[0] to a variable

df = df[0]
print(" df is of type :",type(df))

 df is of type : <class 'pandas.core.frame.DataFrame'>


### Data Analysis

In [7]:
print("sample of the scraped data..")

df.head()

sample of the scraped data..


Postal Code           Borough               Neighborhood
0         M1A      Not assigned                        NaN
1         M2A      Not assigned                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [8]:
# let's see what type of values does Borough contain !!

df['Borough'].value_counts()

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
East Toronto         5
York                 5
Mississauga          1
Name: Borough, dtype: int64

As we can see there are <b>77 rows</b> with value for <b>Borough</b> as <b>Not assigned</b>

In [9]:
# let's see what type of values does Neighbourhood contain !!

result = pd.isna(df['Neighborhood'])

result.value_counts()

False    103
True      77
Name: Neighborhood, dtype: int64

The null values in Neighborhood are same as no of missing values as Borough Column , using this we can say that there isn't a single row that contains empty values in a single column only i.e if we remove the rows with missing values then our datraframe doesn't contain any null values

In [10]:
#counting no of rows that contain Not assigned values for Borough

df.dropna(inplace = True) # inplace parameter help in retaining the dataFrame with valid entries in the same variable.

In [11]:
df.shape

(103, 3)

After removing 77 rows with empty values we have 103 rows remaining

reading the csv file provided http://cocl.us/Geospatial_data

In [12]:
geo_data = pd.read_csv('http://cocl.us/Geospatial_data') #reading csv file from the link

In [13]:
geo_data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [14]:
df_join = pd.merge(df, geo_data, on='Postal Code', how='inner') # merging neighborhood df and location df 

In [15]:
df_join.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

### Use geopy library to get the latitude and longitude values of  Toronto City.

In [16]:
!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1

In [132]:
import json # library to handle JSON files

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

from IPython.core.display import HTML  # to display maps on when publishing to github

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent tnt_explorer, as shown below.

In [57]:
address = 'Toronto , Ontario , CA'

geolocator = Nominatim(user_agent="tnt_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto city are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto city are 43.6534817, -79.3839347.


### Create a map of Toronto with neighborhoods superimposed on top

In [133]:
# create map of Toronto using latitude and longitude values
toronto_map = folium.Map(location=[latitude, longitude], zoom_start=10)


# add markers to map
for lat, lng, borough, neighborhood in zip(df_join['Latitude'], df_join['Longitude'], df_join['Borough'], df_join['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(toronto_map)  
    
HTML(toronto_map._repr_html_())

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Toronto. So let's slice the original dataframe and create a new dataframe of the Toronto data.

In [59]:
toronto_data = df_join[df_join['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
toronto_data.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

Let's get the geographical coordinates of Toronto.

In [119]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="tnt_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [134]:
# create map of Toronto using latitude and longitude values
toronto_map = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(toronto_map)  
    
HTML(toronto_map._repr_html_())

### Define Foursquare Credentials and Version

In [123]:
CLIENT_ID = ' your Foursquare ID'
CLIENT_SECRET = ' your Foursquare Secret'
VERSION = '20200130' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)
print('credentials removed due to privacy concern')

Your credentails:
CLIENT_ID:  your Foursquare ID
CLIENT_SECRET: your Foursquare Secret
credentials removed due to privacy concern


### Let's explore the first neighborhood in our dataframe.

In [86]:
toronto_data.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

In [87]:
toronto_data.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

Get the neighborhood's latitude and longitude values.

In [88]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


### Now, let's get the top 100 venues that are in Regent Park, Harbourfront within a radius of 500 meters.

In [89]:
# type your answer here
latitude = neighborhood_latitude
longitudee = neighborhood_longitude
radius = 500
LIMIT = 100

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude, 
    longitudee, 
    radius, 
    LIMIT)




In [90]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5eb7a0d0d03993001b4617c5'},
 'response': {'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 48,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.653446723052674,
          'lng': -79.3620167174383}],
        'distance': 143,
       

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [91]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [92]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2      Morning Glory Cafe       Breakfast Spot  43.653947 -79.361149
3  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [93]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

48 venues were returned by Foursquare.


### Explore Neighborhoods in Toronto

#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [94]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [95]:
# type your answer here

toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )



Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


### Let's check the size of the resulting dataframe

In [96]:
print(toronto_venues.shape)
toronto_venues.head()

(1228, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

Let's check how many venues were returned for each neighborhood

In [97]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     62   
Christie                                                               17   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              56   
Queen's Park, Ontario Provincial Government                            39   
Regent Park, Harbourfront                                              48   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
St. James Town                                                         77   
St. James Town, Cabbagetown                                            50   
Stn A PO Boxes                                                         95   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      62   
Christie                                                                17   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               56   
Queen's Park, Ontario Provincial Government                             39   
Regent Park, Harbourfront                                               48   
Richmond, Adelaide, King                                                94   
Rosedale                                                                 4   
St. James Town                                                          77   
St. James Town, Cabbagetown                                             50   
Stn A PO Boxes                                                          95   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          35   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            57              57   
CN Tower, King and Spadina, Railway Lands, Harb...     16              16   
Central Bay Street                                     62              62   
Christie                                               17              17   
Church and Wellesley                                   78              78   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

#### Let's find out how many unique categories can be curated from all the returned venues

In [98]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 210 uniques categories.


### Analyze Each Neighborhood

In [99]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Thai Restaurant  Theater  Theme Restaurant  Trail  \
0             0  ...                0        0                 0      0   
1             0  ...                0        0                 0      0   
2             0  ...                0        0                 0      0   
3             0  ...                0        0                 0      0   
4             0  ...                0        0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Women's Store  
0                      0         0              0  
1                      0         0              0  
2                      0         0              0  
3                      0         0              0  
4                      0         0              0  

[5 rows x 210 columns]

In [100]:
toronto_onehot.shape

(1228, 210)

In [101]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
2                                  Central Bay Street     0.016129   
3                                            Christie     0.000000   
4                                Church and Wellesley     0.025641   
5                      Commerce Court, Victoria Hotel     0.000000   
6              First Canadian Place, Underground city     0.000000   
7                            Garden District, Ryerson     0.000000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000   
9           Kensington Market, Chinatown, Grange Park     0.000000   
10        Queen's Park, Ontario Provincial Government     0.025641   
11                          Regent Park, Harbourfront     0.020833   
12                           Richmond, Adelaide, King     0.000000   
13                                           Rosedale     0.000000   
14                                     St. James Town     0.000000   
15                        St. James Town, Cabbagetown     0.000000   
16                                     Stn A PO Boxes     0.010526   
17           Toronto Dominion Centre, Design Exchange     0.000000   
18                     University of Toronto, Harbord     0.028571   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0.000000   0.0000              0.0000        0.0000   
1            0.000000   0.0625              0.0625        0.0625   
2            0.000000   0.0000              0.0000        0.0000   
3            0.000000   0.0000              0.0000        0.0000   
4            0.012821   0.0000              0.0000        0.0000   
5            0.000000   0.0000              0.0000        0.0000   
6            0.000000   0.0000              0.0000        0.0000   
7            0.000000   0.0000              0.0000        0.0000   
8            0.000000   0.0000              0.0000        0.0000   
9            0.000000   0.0000              0.0000        0.0000   
10           0.000000   0.0000              0.0000        0.0000   
11           0.000000   0.0000              0.0000        0.0000   
12           0.000000   0.0000              0.0000        0.0000   
13           0.000000   0.0000              0.0000        0.0000   
14           0.000000   0.0000              0.0000        0.0000   
15           0.000000   0.0000              0.0000        0.0000   
16           0.000000   0.0000              0.0000        0.0000   
17           0.000000   0.0000              0.0000        0.0000   
18           0.000000   0.0000              0.0000        0.0000   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0            0.000            0.000             0.000             0.000000   
1            0.125            0.125             0.125             0.000000   
2            0.000            0.000             0.000             0.000000   
3            0.000            0.000             0.000             0.000000   
4            0.000            0.000             0.000             0.012821   
5            0.000            0.000             0.000             0.040000   
6            0.000            0.000             0.000             0.030000   
7            0.000            0.000             0.000             0.000000   
8            0.000            0.000             0.000             0.000000   
9            0.000            0.000             0.000             0.000000   
10           0.000            0.000             0.000             0.000000   
11           0.000            0.000             0.000             0.000000   
12           0.000            0.000             0.000             0.021277   
13           0.000            0.000             0.000             0.000000   
14           0.000            0.000             0.000             0.038961   
15           0.000            0

#### Let's print each neighborhood along with the top 5 most common venues

In [102]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2            Beer Bar  0.04
3  Seafood Restaurant  0.04
4                Café  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.12
1   Airport Service  0.12
2  Airport Terminal  0.12
3     Boat or Ferry  0.06
4   Harbor / Marina  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.18
1  Italian Restaurant  0.08
2                Café  0.06
3      Sandwich Place  0.05
4     Bubble Tea Shop  0.03


----Christie----
           venue  freq
0  Grocery Store  0.24
1           Café  0.18
2           Park  0.12
3      Nightclub  0.06
4     Restaurant  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.09
1     Sushi Restaurant  0.08
2  Japanese Restaurant  0.06
3           Restaurant  0.04
4        

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [103]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [104]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar            Restaurant              Beer Bar   
1       Airport Service      Airport Terminal      Sculpture Garden   
2    Italian Restaurant                  Café        Sandwich Place   
3                  Café                  Park           Coffee Shop   
4      Sushi Restaurant   Japanese Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery    Seafood Restaurant                  Café   
1         Boat or Ferry               Airport    Airport Food Court   
2           Salad Place        Ice Cream Shop   Japanese Restaurant   
3            Baby Store    Athletics & Sports           Candy Store   
4           Yoga Studio            Smoke Shop               Gay Bar   

  8th Most Common Venue 9th Most Common Venue    10th Most Common Venue  
0           Cheese Shop                   Pub            Breakfast Spot  
1          Airport Gate       Harbor / Marina                     Plane  
2       Thai Restaurant          Burger Joint           Bubble Tea Shop  
3            Restaurant    Italian Restaurant                     Diner  
4                 Hotel             Gastropub  Mediterranean Restaurant

### Cluster Neighborhoods

In [105]:
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 3, 0, 2, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [106]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0                  Café   
4  43.644771 -79.373306               0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                  Park                   Pub   
1      Sushi Restaurant                   Gym                 Diner   
2           Coffee Shop                  Café        Cosmetics Shop   
3           Coffee Shop   American Restaurant             Gastropub   
4          Cocktail Bar            Restaurant              Beer Bar   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot            Restaurant               Theater   
1                  Park    Mexican Restaurant             Juice Bar   
2            Restaurant    Italian Restaurant       Bubble Tea Shop   
3          Cocktail Bar                   Gym              Beer Bar   
4                Bakery    Seafood Restaurant                  Café   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0                  Café                       Bank    Distribution Center  
1   Japanese Restaurant         Italian Restaurant             Hobby Shop  
2   Japanese Restaurant  Middle Eastern Restaurant                Theater  
3            Restaurant         Seafood Restaurant     Italian Restaurant  
4           Cheese Shop                        Pub         Breakfast Spot

Finally, let's visualize the resulting clusters

In [135]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
HTML(map_clusters._repr_html_())

In [111]:
toronto_merged[toronto_merged['Cluster Labels'] == 0 ]

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M5E  Downtown Toronto   
5          M5G  Downtown Toronto   
7          M5H  Downtown Toronto   
8          M5J  Downtown Toronto   
9          M5K  Downtown Toronto   
10         M5L  Downtown Toronto   
11         M5S  Downtown Toronto   
12         M5T  Downtown Toronto   
15         M5W  Downtown Toronto   
16         M4X  Downtown Toronto   
17         M5X  Downtown Toronto   
18         M4Y  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                           Regent Park, Harbourfront  43.654260 -79.360636   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         Berczy Park  43.644771 -79.373306   
5                                  Central Bay Street  43.657952 -79.387383   
7                            Richmond, Adelaide, King  43.650571 -79.384568   
8   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
9            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
10                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
11                     University of Toronto, Harbord  43.662696 -79.400049   
12          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
15                                     Stn A PO Boxes  43.646435 -79.374846   
16                        St. James Town, Cabbagetown  43.667967 -79.367675   
17             First Canadian Place, Underground city  43.648429 -79.382280   
18                               Church and Wellesley  43.665860 -79.383160   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0                0           Coffee Shop                Bakery   
1                0           Coffee Shop      Sushi Restaurant   
2                0        Clothing Store           Coffee Shop   
3                0                  Café           Coffee Shop   
4                0           Coffee Shop          Cocktail Bar   
5                0           Coffee Shop    Italian Restaurant   
7                0           Coffee Shop                  Café   
8                0           Coffee Shop              Aquarium   
9                0           Coffee Shop                  Café   
10               0           Coffee Shop            Restaurant   
11               0                  Café             Bookstore   
12               0                  Café           Coffee Shop   
15               0           Coffee Shop                  Café   
16               0           Coffee Shop            Restaurant   
17               0           Coffee Shop                  Café   
18               0           Coffee Shop      Sushi Restaurant   

   3rd Most Common Venue  4th Most Common Venue 5th Most Common Venue  \
0                   Park                    Pub        Breakfast Spot   
1                    Gym                  Diner                  Park   
2                   Café         Cosmetics Shop            Restaurant   
3    American Restaurant              Gastropub          Cocktail Bar   
4             Restaurant               Beer Bar                Bakery   
5                   Café         Sandwich Place           Salad Place   
7             Restaurant                    Gym       Thai Restaurant   
8                   Café                  Hotel   Sporting Goods Shop   
9                  Hotel             Restaurant   Japanese Restaurant   
10                  Café                  Hotel                   Gym   
11                   Bar     Italian Restaurant   Japanese Restaurant   
12    Mexican Restaurant  Vietnamese Restaurant

As we can see places in <b>cluster 0</b> contains <b>Coffe Shop and cafe</b> as 1st common venue

In [112]:
toronto_merged[toronto_merged['Cluster Labels'] == 1 ]

Postal Code           Borough Neighborhood   Latitude  Longitude  \
14         M4W  Downtown Toronto     Rosedale  43.679563 -79.377529   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
14               1                  Park                 Trail   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
14            Playground         Women's Store              Creperie   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
14      Doner Restaurant               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
14        Discount Store                  Diner

As we can see places in <b>cluster 1</b> contains <b>Park</b> as 1st common venue

In [113]:
toronto_merged[toronto_merged['Cluster Labels'] == 2 ]

Postal Code           Borough Neighborhood   Latitude  Longitude  \
6         M6G  Downtown Toronto     Christie  43.669542 -79.422564   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6               2         Grocery Store                  Café   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6                  Park           Coffee Shop            Baby Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6    Athletics & Sports           Candy Store            Restaurant   

  9th Most Common Venue 10th Most Common Venue  
6    Italian Restaurant                  Diner

As we can see places in <b>cluster 2</b> contains <b>Grocery Store</b> as 1st common venue

In [114]:
toronto_merged[toronto_merged['Cluster Labels'] == 3 ]

Postal Code           Borough  \
13         M5V  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
13  CN Tower, King and Spadina, Railway Lands, Har...  43.628947  -79.39442   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
13               3        Airport Lounge       Airport Service   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
13      Airport Terminal      Sculpture Garden         Boat or Ferry   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
13               Airport    Airport Food Court          Airport Gate   

   9th Most Common Venue 10th Most Common Venue  
13       Harbor / Marina                  Plane

As we can see places in <b>cluster 3</b> contains <b>Airport related services</b> as more common venues